File: ZeroNet-Mobile/app/src/main/java/com/zeronet/mobile/data/Sources.kt
package com.zeronet.mobile.data

import org.json.JSONArray
import org.json.JSONObject

/**
 * Community config feeds on GitHub, tiered by what measurement from inside
 * Iran showed (PLAN.md §3). Tier 0 is this project's own tested list. Tier 1
 * is fetched on every cold connect (when tier 0 was not enough) and is
 * small; tier 2 only when tier 1 did not yield enough working configs; tier 3
 * only when the user asks to search harder.
 */
data class FeedSource(
    val id: String,
    val name: String,
    val repo: String,
    val url: String,
    val tier: Int,
    /** Detached-signature URL (`<url>.sig`); set for the project's own signed list. */
    val sigUrl: String? = null,
) {
    fun toJson(): JSONObject = JSONObject().put("id", id).put("url", url).put("tier", tier)
        .apply { if (sigUrl != null) put("sig_url", sigUrl) }
}

object Sources {
    private const val RAW = "https://raw.githubusercontent.com"

    val builtIn: List<FeedSource> = listOf(
        // Tested every two hours by the harvest workflow and published by
        // this project: only working, encrypted configs, a few hundred at
        // most. Tier 0, so the search tries it before any other feed and,
        // when it yields enough, never downloads the big ones. Served from
        // jsDelivr, which is often reachable when raw.githubusercontent.com
        // is not.
        FeedSource("zeronet", "ZeroNet verified", "zeghostwriter/ZeroNet", "https://cdn.jsdelivr.net/gh/zeghostwriter/ZeroNet@crowd-data/verified.txt", 0, "https://cdn.jsdelivr.net/gh/zeghostwriter/ZeroNet@crowd-data/verified.txt.sig"),
        FeedSource("limilco", "liMilCo", "liMilCo/v2r", "$RAW/liMilCo/v2r/main/new_configs.txt", 1),
        FeedSource("sinavm", "SVM", "sinavm/SVM", "$RAW/sinavm/SVM/main/lite/subscriptions/xray/base64/mix", 1),
        FeedSource("anonymou3", "Multi Proxy (tested)", "4n0nymou3/multi-proxy-config-fetcher", "$RAW/4n0nymou3/multi-proxy-config-fetcher/main/configs/proxy_configs_tested.txt", 1),
        FeedSource("radikal", "0xRadikal", "0xRadikal/Free-v2ray-Configs", "$RAW/0xRadikal/Free-v2ray-Configs/main/all/configs.txt", 2),
        FeedSource("epodonios", "Epodonios", "Epodonios/v2ray-configs", "$RAW/Epodonios/v2ray-configs/main/All_Configs_Sub.txt", 2),
        FeedSource("delta", "Delta-Kronecker", "Delta-Kronecker/V2ray-Config", "$RAW/Delta-Kronecker/V2ray-Config/main/config/all_configs.txt", 3),
        FeedSource("mheidari", "mheidari98", "mheidari98/.proxy", "$RAW/mheidari98/.proxy/main/all", 3),
        FeedSource("f0rc3run", "F0rc3Run", "F0rc3Run/F0rc3Run", "$RAW/F0rc3Run/F0rc3Run/main/Best-Results/sub.txt", 3),
    )

    fun enabled(disabled: Set<String>, maxTier: Int = 3): List<FeedSource> =
        builtIn.filter { it.id !in disabled && it.tier <= maxTier }

    fun toJson(sources: List<FeedSource>): JSONArray = JSONArray().also { a -> sources.forEach { a.put(it.toJson()) } }
}

File: crates/zero-discovery/src/sources.rs
//! The public config feeds a client searches, tiered by what measurement
//! from inside Iran showed. Tier 0 is this project's own tested (and signed)
//! list; tier 1 is fetched on every cold search that tier 0 did not
//! satisfy; tier 2 only when tier 1 fell short; tier 3 only when asked to
//! search harder. The Android app keeps the same list in `data/Sources.kt`.

use crate::feed::FeedSource;

const RAW: &str = "https://raw.githubusercontent.com";

/// A feed with its display name, for hosts that let the user switch feeds.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NamedSource {
    pub name: &'static str,
    pub source: FeedSource,
}

fn source(id: &str, name: &'static str, url: String, tier: u32, sig_url: Option<String>) -> NamedSource {
    NamedSource {
        name,
        source: FeedSource {
            id: id.into(),
            url,
            tier,
            sig_url,
        },
    }
}

/// Every built-in feed.
pub fn builtin() -> Vec<NamedSource> {
    let verified = "https://cdn.jsdelivr.net/gh/zeghostwriter/ZeroNet@crowd-data/verified.txt";
    vec![
        source("zeronet", "ZeroNet verified", verified.into(), 0, Some(format!("{verified}.sig"))),
        source("limilco", "liMilCo", format!("{RAW}/liMilCo/v2r/main/new_configs.txt"), 1, None),
        source("sinavm", "SVM", format!("{RAW}/sinavm/SVM/main/lite/subscriptions/xray/base64/mix"), 1, None),
        source(
            "anonymou3",
            "Multi Proxy (tested)",
            format!("{RAW}/4n0nymou3/multi-proxy-config-fetcher/main/configs/proxy_configs_tested.txt"),
            1,
            None,
        ),
        source("radikal", "0xRadikal", format!("{RAW}/0xRadikal/Free-v2ray-Configs/main/all/configs.txt"), 2, None),
        source("epodonios", "Epodonios", format!("{RAW}/Epodonios/v2ray-configs/main/All_Configs_Sub.txt"), 2, None),
        source("delta", "Delta-Kronecker", format!("{RAW}/Delta-Kronecker/V2ray-Config/main/config/all_configs.txt"), 3, None),
        source("mheidari", "mheidari98", format!("{RAW}/mheidari98/.proxy/main/all"), 3, None),
        source("f0rc3run", "F0rc3Run", format!("{RAW}/F0rc3Run/F0rc3Run/main/Best-Results/sub.txt"), 3, None),
    ]
}

/// The feeds to search: every built-in one up to `max_tier` not in `disabled`.
pub fn enabled(disabled: &[String], max_tier: u32) -> Vec<FeedSource> {
    builtin()
        .into_iter()
        .filter(|named| named.source.tier <= max_tier && !disabled.iter().any(|id| id == &named.source.id))
        .map(|named| named.source)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ids_are_unique_and_the_verified_list_is_signed_and_first() {
        let all = builtin();
        let mut ids: Vec<_> = all.iter().map(|n| n.source.id.clone()).collect();
        ids.sort();
        ids.dedup();
        assert_eq!(ids.len(), all.len());
        assert_eq!(all[0].source.tier, 0);
        assert!(all[0].source.sig_url.is_some());
        assert!(all.iter().all(|n| n.source.url.starts_with("https://")));
    }

    #[test]
    fn disabled_feeds_and_higher_tiers_are_left_out() {
        let some = enabled(&["limilco".into()], 1);
        assert!(some.iter().all(|s| s.tier <= 1 && s.id != "limilco"));
        assert!(some.iter().any(|s| s.id == "zeronet"));
    }
}
